### PR TITLE
second class dropout added

### PR DESCRIPTION
**Pull request by kartik4949**

### CODE REVIEW

**Comment by kartik4949** [comment]
tests?

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -203,6 +203,9 @@ def tanh(self):
   def leakyrelu(self, neg_slope=0.01):
     return self.relu() + (-neg_slope*self).relu()
 
+  def dropout(self, prob=0.5):
+    return self*Tensor(np.asarray(np.random.binomial(1, 1-prob, size=self.shape), dtype=self.dtype), requires_grad=False, gpu=self.gpu)
```
Comment (geohot) [inline tinygrad/tensor.py:207]: So long line! Maybe
```
mask = np.random.binomial(1, 1-prob, size=self.shape).astype(self.dtype)
return self * Tensor(mask, requires_grad=False, gpu=self.gpu)
```

**Comment by geohot** [comment]
Of course it needs a test

**Comment by kartik4949** [comment]
@geohot found interesting thing george
```
template<bool feature_dropout, bool alpha_dropout, bool inplace, typename T>
Ctype<inplace> _dropout_impl(T& input, double p, bool train) {
  TORCH_CHECK(p >= 0 && p <= 1, "dropout probability has to be between 0 and 1, but got ", p);
  if (p == 0 || !train || input.numel() == 0) {
    return input;
  }

  if (p == 1) {
    return multiply<inplace>(input, at::zeros({}, input.options()));
  }

  at::Tensor b; // used for alpha_dropout only
  auto noise = feature_dropout ? make_feature_noise(input) : at::empty_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
  noise.bernoulli_(1 - p);
  if (alpha_dropout) {
    constexpr double alpha = 1.7580993408473766;
    double a = 1. / std::sqrt((alpha * alpha * p + 1) * (1 - p));
    b = noise.add(-1).mul_(alpha * a).add_(alpha * a * p);
    noise.mul_(a);
  } else {
    noise.div_(1 - p);
  }

  if (!alpha_dropout) {
    return multiply<inplace>(input, noise);
  } else {
    return multiply<inplace>(input, noise).add_(b);
  }
}
```

digged pytorch haha
to see there implementation
turns out they are using inverted dropout

so idea of inverted drop is 

normal dropout:
     during training we delete some nodes so in the next layer the value goes is lower (because sum(x1*w1 + x2*w2 + ....))
     but during testing we dont do dropout at all and values are normal... but high for the model because it is used to small values because of dropout

to solve above inverted dropout was introduced

so inverted dropout

mask = noise.bernoulli(1-p)
ret =self * mask
ret = ret/(1-p) # scaling the return values so that in testing when dropout is off we dont have high value as model will now get used to high values becuase of scalling

:) hope this explains

**Comment by kartik4949** [comment]
https://github.com/pytorch/pytorch/blob/22a34bcf4e5eaa348f0117c414c3dd760ec64b13/aten/src/ATen/native/Dropout.cpp

relevant file ... had to dig a lot :(

**Comment by geohot** [comment]
Needs the suggested refactor + getting the multiplication factor right to merge.
